docs/m3db/architecture/engine.md

```diff
@@ -130,75 +109,138 @@ A read begins when an M3DB client calls the [`FetchBatchResult`](https://github.
 
-1. Consult the in-memory bloom filter to determine if its possible the series exists on disk.
-2. If the bloom filter returns positive, then binary search the in-memory index summaries to find the nearest index entry that is *before* the series ID that we're searching for. Review the `index_lookup.go` file for implementation details.
+1. Consult the in-memory bloom filter to determine if it's possible the series exists on disk.
+2. If the bloom filter returns negative, we are sure that the series isn't there, so return that result. If the bloom filter returns positive, then binary search the in-memory index summaries to find the nearest index entry that is *before* the series ID that we're searching for. Review the `index_lookup.go` file for implementation details.
 3. Jump to the offset in the index file that we obtained from the binary search in the previous step, and begin scanning forward until we identify the index entry for the series ID we're looking for *or* we get far enough in the index file that it becomes clear that the ID we're looking for doesn't exist (this is possible because the index file is sorted by ID)
 4. Jump to the offset in the data file that we obtained from scanning the index file in the previous step, and begin streaming data.
 
 Once M3DB has retrieved the three blocks from their respective locations in memory / on-disk, it will transmit all of the data back to the client. Whether or not the client returns a success to the caller for the read is dependent on the configured [consistency level](consistencylevels.md).
 
 **Note:** Since M3DB nodes return compressed blocks (the M3DB client decompresses them), it's not possible to return "partial results" for a given block. If any portion of a read request spans a given block, then that block in its entirety must be transmitted back to the client. In practice, this ends up being not much of an issue because of the high compression ratio that M3DB is able to achieve.
 
+## Buffer
+
+Each series object contains a buffer, which is in charge of handling all data that has yet to be flushed - new writes and bootstrapped data. To accomplish this, it keeps "buckets" of encoders (for new writes) and blocks (for bootstrapped data). M3TSZ, the database's encoding scheme, is designed for compressing time series data in which each datapoint has a timestamp that is larger than the last encoded datapoint. For metrics workloads this works very well because every subsequent datapoint is almost always after the previous one. However, out of order writes will occasionally be received, for example due to clock skew. When this happens, M3DB will allocate a new encoder for the out of order datapoints. These encoders are contained in a bucket along with any blocks that got bootstrapped.
```

Comment:
maybe say: "it keeps mutable "buckets"..." and "immutable blocks (for bootstrapped..."

Reply:
Maybe say instead of M3TSZ is the encoding scheme, say that M3DB has two different encoding schemes. M3TSZ which blah blah blah and Protobuf which blah blah blah and can link to the protobuf encoding docs